File: examples/Java/rtdealer.java
package guide;

import java.util.Random;

import org.zeromq.SocketType;
import org.zeromq.ZMQ;
import org.zeromq.ZMQ.Socket;
import org.zeromq.ZContext;

/**
 * ROUTER-TO-REQ example
 */
public class rtdealer
{
    private static Random    rand        = new Random();
    private static final int NBR_WORKERS = 10;

    private static class Worker extends Thread
    {
        @Override
        public void run()
        {
            try (ZContext context = new ZContext()) {
                Socket worker = context.createSocket(SocketType.DEALER);
                ZHelper.setId(worker); //  Set a printable identity

                worker.connect("tcp://localhost:5671");

                int total = 0;
                while (true) {
                    //  Tell the broker we're ready for work
                    worker.sendMore("");
                    worker.send("Hi Boss");

                    //  Get workload from broker, until finished
                    worker.recvStr(); //  Envelope delimiter
                    String workload = worker.recvStr();
                    boolean finished = workload.equals("Fired!");
                    if (finished) {
                        System.out.printf("Completed: %d tasks\n", total);
                        break;
                    }
                    total++;

                    //  Do some random work
                    try {
                        Thread.sleep(rand.nextInt(500) + 1);
                    }
                    catch (InterruptedException e) {
                    }
                }
            }
        }
    }

    /**
     * While this example runs in a single process, that is just to make
     * it easier to start and stop the example. Each thread has its own
     * context and conceptually acts as a separate process.
     */
    public static void main(String[] args) throws Exception
    {
        try (ZContext context = new ZContext()) {
            Socket broker = context.createSocket(SocketType.ROUTER);
            broker.bind("tcp://*:5671");

            for (int workerNbr = 0; workerNbr < NBR_WORKERS; workerNbr++) {
                Thread worker = new Worker();
                worker.start();
            }

            //  Run for five seconds and then tell workers to end
            long endTime = System.currentTimeMillis() + 5000;
            int workersFired = 0;
            while (true) {
                //  Next message gives us least recently used worker
                String identity = broker.recvStr();
                broker.sendMore(identity);
                broker.recv(0); //  Envelope delimiter
                broker.recv(0); //  Response from worker
                broker.sendMore("");

                //  Encourage workers until it's time to fire them
                if (System.currentTimeMillis() < endTime)
                    broker.send("Work harder");
                else {
                    broker.send("Fired!");
                    if (++workersFired == NBR_WORKERS)
                        break;
                }
            }
        }
    }
}

File: examples/Go/rtdealer.go
//
// ROUTER-to-DEALER example
//
package main

import (
	"fmt"
	zmq "github.com/alecthomas/gozmq"
	"math/rand"
	"strings"
	"time"
)

const NBR_WORKERS int = 10

func randomString() string {
	source := "abcdefghijklmnopqrstuvwxyz"
	target := make([]string, 20)
	for i := 0; i < 20; i++ {
		target[i] = string(source[rand.Intn(len(source))])
	}
	return strings.Join(target, "")
}

func worker_task() {
	context, _ := zmq.NewContext()
	defer context.Close()

	worker, _ := context.NewSocket(zmq.DEALER)
	defer worker.Close()
	worker.SetIdentity(randomString())
	worker.Connect("tcp://localhost:5671")

	total := 0
	for {
		//  Tell the broker we're ready for work
		worker.SendMultipart([][]byte{[]byte(""), []byte("Hi Boss")}, 0)

		//  Get workload from broker, until finished
		parts, _ := worker.RecvMultipart(0)
		workload := parts[1]
		if string(workload) == "Fired!" {
			id, _ := worker.Identity()
			fmt.Printf("Completed: %d tasks (%s)\n", total, id)
			break
		}
		total++

		//  Do some random work
		time.Sleep(time.Duration(rand.Intn(500)) * time.Millisecond)
	}
}

//  While this example runs in a single process, that is just to make
//  it easier to start and stop the example. Each thread has its own
//  context and conceptually acts as a separate process.

func main() {
	context, _ := zmq.NewContext()
	defer context.Close()

	broker, _ := context.NewSocket(zmq.ROUTER)
	defer broker.Close()
	broker.Bind("tcp://*:5671")

	rand.Seed(time.Now().Unix())

	for i := 0; i < NBR_WORKERS; i++ {
		go worker_task()
	}

	end_time := time.Now().Unix() + 5
	workers_fired := 0

	for {
		//  Next message gives us least recently used worker
		parts, err := broker.RecvMultipart(0)
		if err != nil {
			print(err)
		}
		identity := parts[0]
		now := time.Now().Unix()
		if now < end_time {
			broker.SendMultipart([][]byte{identity, []byte(""), []byte("Work harder")}, 0)
		} else {
			broker.SendMultipart([][]byte{identity, []byte(""), []byte("Fired!")}, 0)
			workers_fired++
			if workers_fired == NBR_WORKERS {
				break
			}
		}
	}
}
